File: app/lib/models/quesolimpia_parameters.dart
/// Parameters for the QuesoLimpia pass.
/// Professional film dust, dirt and spot restoration.
class QuesoLimpiaParameters {
  final bool enabled;
  final String mode;
  final int strength;
  final int threshold;
  final int spatialThreshold;
  final int minDustSize;
  final int maxDustSize;
  final bool detectBright;
  final bool detectDark;
  final bool detectSpatial;
  final bool detectStatic;
  final int grainSuppress;
  final int edgeProtect;
  final bool sceneProtect;
  final double sceneThreshold;
  final bool chroma;
  final int grainRestore;
  final int temporalRadius;
  final int blksize;
  final int pel;
  final String showMask;

  const QuesoLimpiaParameters({
    this.enabled = false,
    this.mode = 'balanced',
    this.strength = 75,
    this.threshold = 20,
    this.spatialThreshold = 15,
    this.minDustSize = 1,
    this.maxDustSize = 16,
    this.detectBright = true,
    this.detectDark = true,
    this.detectSpatial = true,
    this.detectStatic = true,
    this.grainSuppress = 40,
    this.edgeProtect = 50,
    this.sceneProtect = true,
    this.sceneThreshold = 0.10,
    this.chroma = true,
    this.grainRestore = 30,
    this.temporalRadius = 1,
    this.blksize = 16,
    this.pel = 2,
    this.showMask = 'off',
  });

  QuesoLimpiaParameters copyWith({
    bool? enabled,
    String? mode,
    int? strength,
    int? threshold,
    int? spatialThreshold,
    int? minDustSize,
    int? maxDustSize,
    bool? detectBright,
    bool? detectDark,
    bool? detectSpatial,
    bool? detectStatic,
    int? grainSuppress,
    int? edgeProtect,
    bool? sceneProtect,
    double? sceneThreshold,
    bool? chroma,
    int? grainRestore,
    int? temporalRadius,
    int? blksize,
    int? pel,
    String? showMask,
  }) {
    return QuesoLimpiaParameters(
      enabled: enabled ?? this.enabled,
      mode: mode ?? this.mode,
      strength: strength ?? this.strength,
      threshold: threshold ?? this.threshold,
      spatialThreshold: spatialThreshold ?? this.spatialThreshold,
      minDustSize: minDustSize ?? this.minDustSize,
      maxDustSize: maxDustSize ?? this.maxDustSize,
      detectBright: detectBright ?? this.detectBright,
      detectDark: detectDark ?? this.detectDark,
      detectSpatial: detectSpatial ?? this.detectSpatial,
      detectStatic: detectStatic ?? this.detectStatic,
      grainSuppress: grainSuppress ?? this.grainSuppress,
      edgeProtect: edgeProtect ?? this.edgeProtect,
      sceneProtect: sceneProtect ?? this.sceneProtect,
      sceneThreshold: sceneThreshold ?? this.sceneThreshold,
      chroma: chroma ?? this.chroma,
      grainRestore: grainRestore ?? this.grainRestore,
      temporalRadius: temporalRadius ?? this.temporalRadius,
      blksize: blksize ?? this.blksize,
      pel: pel ?? this.pel,
      showMask: showMask ?? this.showMask,
    );
  }

  /// Get a summary string for display.
  String get summary {
    if (!enabled) return 'Off';
    final parts = <String>[];
    parts.add(mode[0].toUpperCase() + mode.substring(1));
    if (detectStatic) parts.add('Gate Hair');
    if (grainRestore > 0) parts.add('Grain ($grainRestore%)');
    return 'QuesoLimpia (${parts.join(", ")})';
  }

  factory QuesoLimpiaParameters.fromJson(Map<String, dynamic> json) {
    return QuesoLimpiaParameters(
      enabled: json['enabled'] as bool? ?? false,
      mode: json['mode'] as String? ?? 'balanced',
      strength: json['strength'] as int? ?? 75,
      threshold: json['threshold'] as int? ?? 20,
      spatialThreshold: json['spatialThreshold'] as int? ?? 15,
      minDustSize: json['minDustSize'] as int? ?? 1,
      maxDustSize: json['maxDustSize'] as int? ?? 16,
      detectBright: json['detectBright'] as bool? ?? true,
      detectDark: json['detectDark'] as bool? ?? true,
      detectSpatial: json['detectSpatial'] as bool? ?? true,
      detectStatic: json['detectStatic'] as bool? ?? true,
      grainSuppress: json['grainSuppress'] as int? ?? 40,
      edgeProtect: json['edgeProtect'] as int? ?? 50,
      sceneProtect: json['sceneProtect'] as bool? ?? true,
      sceneThreshold: (json['sceneThreshold'] as num?)?.toDouble() ?? 0.10,
      chroma: json['chroma'] as bool? ?? true,
      grainRestore: json['grainRestore'] as int? ?? 30,
      temporalRadius: json['temporalRadius'] as int? ?? 1,
      blksize: json['blksize'] as int? ?? 16,
      pel: json['pel'] as int? ?? 2,
      showMask: json['showMask'] as String? ?? 'off',
    );
  }

  Map<String, dynamic> toJson() => {
    'enabled': enabled,
    'mode': mode,
    'strength': strength,
    'threshold': threshold,
    'spatialThreshold': spatialThreshold,
    'minDustSize': minDustSize,
    'maxDustSize': maxDustSize,
    'detectBright': detectBright,
    'detectDark': detectDark,
    'detectSpatial': detectSpatial,
    'detectStatic': detectStatic,
    'grainSuppress': grainSuppress,
    'edgeProtect': edgeProtect,
    'sceneProtect': sceneProtect,
    'sceneThreshold': sceneThreshold,
    'chroma': chroma,
    'grainRestore': grainRestore,
    'temporalRadius': temporalRadius,
    'blksize': blksize,
    'pel': pel,
    'showMask': showMask,
  };
}

File: worker/src/models/quesolimpia_parameters.rs
use serde::{Deserialize, Serialize};

/// Parameters for the QuesoLimpia pass.
/// Professional film dust, dirt and spot restoration with motion-aware spatio-temporal detection.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QuesoLimpiaParameters {
    /// Whether this pass is enabled.
    #[serde(default)]
    pub enabled: bool,

    /// Restoration mode preset: "gentle", "balanced", "aggressive", "forensic"
    #[serde(default = "default_mode")]
    pub mode: String,

    /// General strength scale (10-100, default 75)
    #[serde(default = "default_strength")]
    pub strength: i32,

    /// Temporal base threshold (1-60, default 20)
    #[serde(default = "default_threshold")]
    pub threshold: i32,

    /// Spatial threshold for low-contrast dirt (1-40, default 15)
    #[serde(default = "default_spatial_threshold")]
    pub spatial_threshold: i32,

    /// Minimum dust spot size in pixels (0-8, default 1)
    #[serde(default = "default_min_dust_size")]
    pub min_dust_size: i32,

    /// Maximum dust spot size in pixels (2-64, default 16)
    #[serde(default = "default_max_dust_size")]
    pub max_dust_size: i32,

    /// Detect bright dust spots (default true)
    #[serde(default = "default_true")]
    pub detect_bright: bool,

    /// Detect dark dust spots (default true)
    #[serde(default = "default_true")]
    pub detect_dark: bool,

    /// Detect low-contrast spatial spots (default true)
    #[serde(default = "default_true")]
    pub detect_spatial: bool,

    /// Detect static defects / gate hair (default true)
    #[serde(default = "default_true")]
    pub detect_static: bool,

    /// Grain suppression to avoid false positives (0-100, default 40)
    #[serde(default = "default_grain_suppress")]
    pub grain_suppress: i32,

    /// Edge protection with Canny (0-100, default 50)
    #[serde(default = "default_edge_protect")]
    pub edge_protect: i32,

    /// Scene cut protection (default true)
    #[serde(default = "default_true")]
    pub scene_protect: bool,

    /// Scene cut sensitivity threshold (default 0.10)
    #[serde(default = "default_scene_threshold")]
    pub scene_threshold: f64,

    /// Process chroma planes (default true)
    #[serde(default = "default_true")]
    pub chroma: bool,

    /// Grain compensation in repaired areas (0-100, default 30)
    #[serde(default = "default_grain_restore")]
    pub grain_restore: i32,

    /// Temporal radius: 1=±1 frame, 2=±2 frames (default 1)
    #[serde(default = "default_temporal_radius")]
    pub temporal_radius: i32,

    /// Block size for motion analysis (default 16)
    #[serde(default = "default_blksize")]
    pub blksize: i32,

    /// Sub-pixel accuracy (default 2)
    #[serde(default = "default_pel")]
    pub pel: i32,

    /// Diagnostic visualization mode: "off", "raw", "refined", "repair", "static", "side_by_side"
    #[serde(default = "default_show_mask")]
    pub show_mask: String,
}

fn default_true() -> bool { true }
fn default_mode() -> String { "balanced".to_string() }
fn default_strength() -> i32 { 75 }
fn default_threshold() -> i32 { 20 }
fn default_spatial_threshold() -> i32 { 15 }
fn default_min_dust_size() -> i32 { 1 }
fn default_max_dust_size() -> i32 { 16 }
fn default_grain_suppress() -> i32 { 40 }
fn default_edge_protect() -> i32 { 50 }
fn default_scene_threshold() -> f64 { 0.10 }
fn default_grain_restore() -> i32 { 30 }
fn default_temporal_radius() -> i32 { 1 }
fn default_blksize() -> i32 { 16 }
fn default_pel() -> i32 { 2 }
fn default_show_mask() -> String { "off".to_string() }

impl Default for QuesoLimpiaParameters {
    fn default() -> Self {
        Self {
            enabled: false,
            mode: default_mode(),
            strength: default_strength(),
            threshold: default_threshold(),
            spatial_threshold: default_spatial_threshold(),
            min_dust_size: default_min_dust_size(),
            max_dust_size: default_max_dust_size(),
            detect_bright: true,
            detect_dark: true,
            detect_spatial: true,
            detect_static: true,
            grain_suppress: default_grain_suppress(),
            edge_protect: default_edge_protect(),
            scene_protect: true,
            scene_threshold: default_scene_threshold(),
            chroma: true,
            grain_restore: default_grain_restore(),
            temporal_radius: default_temporal_radius(),
            blksize: default_blksize(),
            pel: default_pel(),
            show_mask: default_show_mask(),
        }
    }
}
